### PR TITLE
Proposal: add $deprecated property

### PR DESCRIPTION
**Pull request by drwpow**
## Summary

This adds a new `$deprecated` property for both tokens and groups [as suggested by @romainmenke](https://github.com/design-tokens/community-group/issues/118).

## Reasoning

This PR basically accepts #118 as-proposed (the original proposal), while taking into account feedback from followup comments.

### Pros

- This mirrors prior art like [JSDoc deprecation](https://jsdoc.app/tags-deprecated) that is well-known and well-used
- Style dictionary also [has this concept](https://github.com/design-tokens/community-group/issues/118#issuecomment-2277926212) and it makes sense to codify in the spec
- This seemed to have fairly-universal support (only disagreements were minor syntax differences, but that also spawned [a separate proposal](https://github.com/design-tokens/community-group/issues/161) that can be evaluated independently)

### Cons

- Expanding `$deprecated` in the future would be a breaking change
- What constitutes “deprecation” is a little ambiguous in the spec as-written. But the hope is that it’s no more ambiguous than using [this exact term elsewhere](https://jsdoc.app/tags-deprecated).
- What “deprecation” means is up to tools to decide (e.g. whether or not to generate output from those tokens, and/or throw warnings/errors)

## Alternatives

- Adding a pair property (e.g. `$deprecated` / `$deprecatedMessage`) isn’t a design pattern that currently exists in the spec. Pair properties can also be antipatterns as they introduce validation complexity (e.g. what if `$deprecatedMessage` is set but `$deprecated` isn’t—is that valid or invalid? and if valid, what behavior is inferred?)
- The object syntax (e.g. `$deprecated.deprecated` / `$deprecated.message`) wasn’t proposed here because it adds unnecessary boilerplate.
- A more complex [status object](https://github.com/design-tokens/community-group/issues/161) would be a separate proposal because that encroaches on opinionated versioning (and versioning won’t map 1:1 to existing tokens)
  - This proposal assumes that any implementation of status/versioning won’t conflict with `$deprecated`. Or in areas where they potentially overlap, will be complementary rather than conflicting.
  - In other words, acceptance of this does NOT mean refusal of the Status Object proposal

## Notes
- This assumes **aliases may be deprecated too** (even if the tokens they point to don’t). Should this be outlined explicitly?
- It is the author’s opinion that deprecation is not directly related to versioning. Versioning is a more complex opinion on what constitutes “breaking changes” from non-, e.g. [semver](https://semver.org/). Though versioning _may_ include deprecation and/or removal of tokens, it’s so much more. Conversely, the act of deprecation is common enough it affects every design system, even those that may not have defined versioning strategies. So whether or not this spec defines an opinionated versioning strategy, shouldn’t affect this proposal.

### CODE REVIEW

File: technical-reports/format/design-token.md [hunk 1]
```diff
@@ -179,4 +179,35 @@ Tool vendors are encouraged to publicly share specifications of their extension
   The extensions section is not limited to vendors. All token users can add additional data in this section for their own purposes.
 </p>
 
-## More token properties TBC
+### Deprecated
+
+The optional **`$deprecated`** property is a boolean or string where tooks MAY specify deprecated tokens. A token MAY be marked deprecated in any of the following scenarios:
```
Comment (drwpow) [inline technical-reports/format/design-token.md:184]: cc @c1rrus any nitpicking of language here would be greatly appreciated 🙏. Happy to rephrase this to match the existing language (or if you wanted to do a pass on it that would be fine, too!)

**Review by romainmenke — APPROVED** [review]
Thank you. LGTM 👍

File: technical-reports/format/design-token.md (hunk 1 above)
Comment (c1rrus) [inline technical-reports/format/design-token.md:184]: Minor typo

```suggestion
The optional **`$deprecated`** property is a boolean or string where tools MAY specify deprecated tokens. A token MAY be marked deprecated in any of the following scenarios:
```

File: technical-reports/format/groups.md [hunk 2]
```diff
@@ -206,6 +206,36 @@ In this example, the "brand" group has 2 extensions: `org.example.tool-a` and `o
 
 </aside>
 
+### Deprecated
+
+The optional **`$deprecated`** property is a boolean or string where tooks MAY specify all tokens within the group are deprecated. A token MAY be marked deprecated in any of the following scenarios:
```
Comment (c1rrus) [inline technical-reports/format/groups.md:211]: Same typo here.

```suggestion
The optional **`$deprecated`** property is a boolean or string where tools MAY specify all tokens within the group are deprecated. A token MAY be marked deprecated in any of the following scenarios:
```

(aside: Maybe we should consider having a single chapter for all token and/or group properties, since many exist for both. Each property could specify if it can be used on tokens, groups or both. That could avoid a lot of the duplication and copy-pasting we need to do today. Obvs. that would need to be a separate PR)

**Review by c1rrus — CHANGES_REQUESTED** [review]
Spotted a couple of teeny typos. But otherwise this looks good. Happy to approve once they get fixed.

File: technical-reports/format/groups.md (hunk 2 above)
Comment (drwpow) [reply technical-reports/format/groups.md]: >(aside: Maybe we should consider having a single chapter for all token and/or group properties, since many exist for both.

That would make sense to me. It could also possibly clarify [some in-the-weeds confusion around parsing token vs group nodes](https://github.com/design-tokens/community-group/issues/253)

**Review by c1rrus — APPROVED** [review]
Nice one. Thanks for soing this @drwpow!

**Comment by drwpow** [comment]
Thanks! Will give other editors a chance to review and leave feedback over the next week or 2. If there’s no objection, we’ll merge!

But also, for the other proposals moving forward, generally the editors express the same opinions found in the comments, so the fact it’s being proposed in the first place is usually positive 🙂

File: technical-reports/format/design-token.md [hunk 3]
```diff
@@ -178,3 +178,36 @@ Tool vendors are encouraged to publicly share specifications of their extension
...
+{
+  "Button background": {
+    "$value": "#777777",
+    "$type": "color",
+    "$deprecated": true
+  },
+  "Button focus": {
+    "$value": "#70c0ff",
+    "$type": "color",
+    "$deprecated": "Please use {button.activeBorder} instead."
```
Comment (CITguy) [inline technical-reports/format/design-token.md:202]: Would downstream tooling need to convert `{button.activeBorder}` to a language-specific name?  This syntax is typically used for aliasing, and there's no clear instruction on:

1. Should this alias be resolved?
    1.  If so, what value should it resolve to?

Comment (drwpow) [reply technical-reports/format/design-token.md:202]: Yeah that’s a good question. I could see this being converted to, say, JSDoc or a CSS comment like so:

```tsx
  /** @deprecated Please use {button.activeBorder} instead */
  'Button focus': '#70c0ff',
```

In which case, would that be clear to people? The simplest interpretation is “the designers/dev write this message in a way that is clearest to their team” with no official guidance. I’m generally against parsing strings in any way, but yeah to your point there is a good question of “could this be improved?”

Nothing comes to mind, because in many cases there is no 1:1 replacement; it’s often a mixture of things. Perhaps we just adjust this example to not mention other tokens to avoid confusion?

File: technical-reports/format/groups.md [hunk 4]
```diff
@@ -206,6 +206,36 @@ In this example, the "brand" group has 2 extensions: `org.example.tool-a` and `o
...
+- A future update to the design system will remove this token
+- Removing the token now may break existing support
+- This token is discouraged from future use
+
+<aside class="example">
+
+```json
+{
+  "Button": {
+    "$deprecated": "Please use {action.*} tokens instead.",
```
Comment (CITguy) [inline technical-reports/format/groups.md:222]: (see [similar comment](https://github.com/design-tokens/community-group/pull/255/files#r1819400982))

File: technical-reports/format/groups.md [hunk 5]
```diff
@@ -206,6 +206,36 @@ In this example, the "brand" group has 2 extensions: `org.example.tool-a` and `o
...
+```
+
+</aside>
+
+In the context of a group, adding `$deprecated` will apply to all tokens within that group, unless a token explicitly sets a value of `false`. Any value provided by a token will override the group default.
+
+| Value    | Explanation                                                 |
+| :------- | :---------------------------------------------------------- |
+| `true`   | This token is deprecated (no explanation provided).         |
+| `String` | This token is deprecated AND this is an explanation.        |
```
Comment (kaelig) [inline technical-reports/format/groups.md:236]: Perhaps it would be useful to mention that tool makers MAY augment the string when it contains aliases such as the one given as an example:

```
Please use {button.activeBorder} instead
```

**Comment by lukasoppermann** [comment]
@drwpow good to merge?

**Comment by drwpow** [comment]
> @drwpow good to merge?

Almost! We had been waiting to see if one group had feedback. Will ping them again and if not, will merge!

**Comment by drwpow** [comment]
> To confirm, there is neither a necessity nor a spec-defined way to confirm or extract an alternate token suggestion inside the deprecated tag, toolmakers can make their own assumptions on if and how to programmatically determine it?

Correct. Updated the wording per @kaelig’s suggestion to describe this a bit more. By default we don’t have aliases in the examples anymore outside the section that describes “it’s up to tools to determine how they do/don’t handle parsing the `$deprecated` string.

**Comment by drwpow** [comment]
We’ve waited for additional parties to give input / feedback but haven’t gotten any pushback or corrections. Merging!
